src/server/routes/oauth.ts

```diff
@@ -82,6 +84,55 @@ const redirectForGenericError = (_: Request, res: ResponseWithRequestState) => {
 	);
 };
 
+/**
+ * A recursive function which waits for the legacy_identity_id claim
+ * to be present in the access token during a succession of token
+ * refreshes.
+ */
+const waitForLegacyIdentityId = async (
+	openIdClient: OpenIdClient,
+	tokenSet: TokenSet,
+	refreshToken: string,
+	maximumWaitTime = 30000,
+	waitTime = 500,
+): Promise<TokenSet | null> => {
+	const { legacy_identity_id } = tokenSet.claims();
+
+	if (legacy_identity_id) {
+		// The legacy_identity_id claim is present, we can return the token set
+		return tokenSet;
+	}
+
+	if (maximumWaitTime <= 0) {
+		// We've timed out waiting for the claim to be present
+		return null;
+	}
+
+	const start = Date.now();
+	const refreshedTokenSet = await openIdClient.refresh(refreshToken);
+	const newRefreshToken = refreshedTokenSet?.refresh_token;
+	if (!newRefreshToken) {
+		// There's no refresh token in this token set, so we can't continue
+		return null;
+	}
+
+	// Non-blocking wait for the next iteration
+	return new Promise((resolve) => {
+		setTimeout(async () => {
+			const remainingTime = maximumWaitTime - (Date.now() - start);
+			resolve(
+				await waitForLegacyIdentityId(
+					openIdClient,
+					refreshedTokenSet,
+					newRefreshToken,
+					remainingTime,
+					waitTime,
+				),
```

Comment:
Could potentially have a slightly simpler nested function that doesn't need to pass the arguments that never change, ie. openIdClient and waitTime.  `refreshedTokenSet` is possibly redundant here I think.